utils: chunked_vector: optimize for trivially_copyable types

bhalevy opened the pull request:
When the chunked_vector holds trivially copyable type T it's more efficient to use std::memcpy to copy or move items around, then to call to copy or move constructors per item.

This change has no significant change on perf-simple-query: 
```
base: median 95954.90 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   42312 insns/op,        0 errors)
post: median 95543.43 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   42343 insns/op,        0 errors)
```

- [x] ** Backport reason (please explain below if this patch should be backported or not) **
optimization, no need to backport

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8701/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8701/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9527/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1504/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8701/)
#### Build Details:
- Duration: 2 hr 49 min
- Builder: spider7.cloudius-systems.com

bhalevy commented:
@scylladb/scylla-maint please consider merging

raphaelsc commented:
> When the chunked_vector holds trivially copyable type T it's more efficient to use std::memcpy to copy or move items around, then to call to copy or move constructors per item.
> 
> perf-simple-query results show slight improvement in instructions/ops:
> 
> ```
> base: median 97702.67 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   42411 insns/op,        0 errors)
> post: median 95048.61 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   42291 insns/op,        0 errors)
> ```
> 

why is tps lower?

bhalevy commented:
> > When the chunked_vector holds trivially copyable type T it's more efficient to use std::memcpy to copy or move items around, then to call to copy or move constructors per item.
> > perf-simple-query results show slight improvement in instructions/ops:
> > ```
> > base: median 97702.67 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   42411 insns/op,        0 errors)
> > post: median 95048.61 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   42291 insns/op,        0 errors)
> > ```
> 
> why is tps lower?

I don't know. It's not very reliable since it's affected by real time, i.e. by the cpu clock and possibly by other processes, so it varies a lot in each run and between runs.
Instructions per op is much more reliable metric.

avikivity commented on utils/chunked_vector.hh:
```diff
@@ -325,8 +327,32 @@ size_t chunked_vector<T, max_contiguous_allocation>::external_memory_usage() con
+            for (auto end = src + n; src != end; ++dst, ++src) {
+                new (dst) T(*src);
+            }
+        }
```
This is std::uninitialized_copy() (which includes the optimization for trivially copyable types).

bhalevy replied:
done in v2

avikivity commented on utils/chunked_vector.hh:
```diff
@@ -397,6 +423,10 @@ chunked_vector<T, max_contiguous_allocation>::new_chunk(size_t n) {
+    if constexpr (std::is_trivially_move_constructible_v<T> && std::is_trivially_destructible_v<T>) {
+        std::memcpy(result, begin, reinterpret_cast<std::uintptr_t>(end) - reinterpret_cast<std::uintptr_t>(begin));
+        return;
+    }
```
Here, we can use std::uninitialized_move() followed by std::destroy().

bhalevy replied:
done in v2

bhalevy commented:
v2 ([f7a15be](https://github.com/scylladb/scylladb/commit/f7a15be14572da390ec81f3784eaf3052038ba84)): use `std::unitinialized_{copy,move}` and `std::destroy`

avikivity commented on utils/chunked_vector.hh:
```diff
@@ -8,6 +8,8 @@
 #pragma once
 
+#include <cstring>
+
```
No longer needed.

bhalevy replied:
dropped in v3

avikivity commented on utils/chunked_vector.hh:
```diff
@@ -325,8 +327,25 @@ size_t chunked_vector<T, max_contiguous_allocation>::external_memory_usage() con
+    auto size = x.size();
+    if (!size) {
+        return;
+    }
```
Probably better to avoid. Most vectors will have content, and the rest of the code will work with zero size.

bhalevy replied:
ok

bhalevy replied:
dropped in v3

avikivity commented on utils/chunked_vector.hh:
```diff
@@ -325,8 +327,25 @@ size_t chunked_vector<T, max_contiguous_allocation>::external_memory_usage() con
+    }
+    if (auto remainder = size % max_chunk_capacity()) {
+        copy_chunk(i, remainder);
+    }
```
The if() is likely unneeded. But maybe it is - can _chunks be empty?

bhalevy replied:
yes, _chunks can be empty if the chunked_vector is empty.
Also _size might be aligned on `max_chunk_capacity()` and then there is no remainder.
We can turn this into an iteration on _chunks.size(), calculating size for chunk for each iteration.
It might be a bit more efficient for single chunk vectors

avikivity replied:
Then copy_chunk(0) will be illegal.

bhalevy replied:
That's why I had the first `if (!size)` in the beginning of the function.
In any case, godbolt .org indicates that a single loop will produce slightly more condensed machine code for x86_64, so I simplified the function in v3, by having a single loop and calculating the count per chunk in the loop body using std::min(size, max_chunk_size())

avikivity replied:
Yes the new loop looks much better.

avikivity replied:
But: it's not exception safe. You must update _size after every uninitialized_copy_n() succeeds, so the destructor can tear it down if the next one fails (it will destroy partial work that it does, not not work that previous iterations did).

avikivity replied:
No, you need a try/catch, since the destructor won't be called on a not-fully-constructed object.

bhalevy replied:
good point, will fix.

bhalevy replied:
Actually, @avikivity, the copy constructor calls the default constructor (which is implicitly noexcept).
So the destructor will called if an exception is thrown in the copy constructor function body, since it's already considered as fully constructed.

bhalevy replied:
In the process, I've found out that the fill constructor isn't safe.  Submitted https://github.com/scylladb/scylladb/pull/18636 (separately, to consider backport)

bhalevy commented:
In v3 ([31ff25a](https://github.com/scylladb/scylladb/commit/31ff25aee33f2e9b3acd94e8d46cd135a7039aab)):
- removed extraneous #include
- simplified copy logic.

avikivity commented on utils/chunked_vector.hh:
```diff
@@ -378,8 +386,13 @@ chunked_vector<T, max_contiguous_allocation>::operator=(chunked_vector&& x) noex
+            std::destroy_n(_chunks[i].get(), max_chunk_capacity());
+        }
+        if (auto remainder = _size % max_chunk_capacity()) {
+            std::destroy_n(_chunks[i].get(), remainder);
```
Can you use a similar nice loop here?

bhalevy replied:
yes, of course, sorry for missing that.

bhalevy replied:
done in v4

avikivity commented on utils/chunked_vector.hh:
```diff
@@ -397,12 +410,8 @@ chunked_vector<T, max_contiguous_allocation>::new_chunk(size_t n) {
-        ++result;
-    }
+    std::uninitialized_move(begin, end, result);
+    std::destroy(begin, end);
```
This is safe, since we require that the move constructor be non-throwing.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8735/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8735/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1534/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9556/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8735/)
#### Build Details:
- Duration: 5 hr 59 min
- Builder: i-0279a9a1ff7985ba8 (m5ad.8xlarge)

bhalevy commented:
In v4 ([2dc42c6](https://github.com/scylladb/scylladb/commit/2dc42c6415e85ceeb280294a6f1b4be80a770ada)):
- rebased
- incrementally update _size in copy ctor to facilitate destruction of partially copied object, on exception.
- simplified destruction loop

bhalevy commented:
Draft until https://github.com/scylladb/scylladb/pull/18636

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8746/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8746/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1543/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9565/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8746/)
#### Build Details:
- Duration: 5 hr 54 min
- Builder: i-04e80069e8e21a6b5 (m5ad.8xlarge)

bhalevy commented:
In v5 ([744c794](https://github.com/scylladb/scylladb/commit/744c794924b1dce041036064c5b2bda8334f3e9c)):
- rebased
- converted also fill ctor to work per-chunk and use std::uninitialized_fill_n

nyh commented:
> When the chunked_vector holds trivially copyable type T it's more efficient to use std::memcpy to copy or move items around, then to call to copy or move constructors per item.
> 
> This change has no significant change on perf-simple-query:
> 
> ```
> base: median 95954.90 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   42312 insns/op,        0 errors)
> post: median 95543.43 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   42343 insns/op,        0 errors)
> ```
> 
>     * [x]  ** Backport reason (please explain below if this patch should be backported or not) **
>       optimization, no need to backport

If this "optimization" doesn't optimize that use case, I wonder what was your original motivation for doing it. Was it something that you thought of while reviewing the code? Do you have a more specific use case in mind which did get optimized by this change? Did you try any sort of unit-test micro-benchmark that got improved by your change?

nyh reviewed (approved):
I looked at the algorithms carefully, and they look correct, and I'm beginning to understand the motivation. In Seastar's chunked_fifo.hh I also had a sophisticated destructor but then realized that it isn't common to destroy a full queue (one normally consumes the queue and not deletes it while it's very full)  so I ifdefed out all the sophistication and left a simple loop implementation.
But I agree that chunked_vector isn't just about a queue - often it holds data that is read without popping - and then the entire vector is destroyed while still full. So this optimization is worthwhile.

nyh commented:
@avikivity can you please leave your final stamp of approval? You had a lot of comments and requests previously.

bhalevy commented:
> > When the chunked_vector holds trivially copyable type T it's more efficient to use std::memcpy to copy or move items around, then to call to copy or move constructors per item.
> > This change has no significant change on perf-simple-query:
> > ```
> > base: median 95954.90 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   42312 insns/op,        0 errors)
> > post: median 95543.43 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   42343 insns/op,        0 errors)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > * [x]  ** Backport reason (please explain below if this patch should be backported or not) **
> >   optimization, no need to backport
> > ```
> 
> If this "optimization" doesn't optimize that use case, I wonder what was your original motivation for doing it. Was it something that you thought of while reviewing the code?

Yes, we should prefer bulk, contiguous operations that can work sequentially on each chunk, over doing per item operations that need to calculate the address for each item, plus they are optimized for trivial types, that we don't currently happen to use on the hot data path.

> Do you have a more specific use case in mind which did get optimized by this change? Did you try any sort of unit-test micro-benchmark that got improved by your change?

https://github.com/scylladb/scylladb/pull/18545 will increase the usage of chunked_vector (with a non trivial type) on the hot path, so I expect to see somw benefits there (but not in perf simple query, since it probably won't generate long enough partition range vectors, maybe we need a separate benchmark for that).
